core/templates/pages/exploration-editor-page/exploration-save-and-publish-buttons/exploration-save-and-publish-buttons.component.ts

```diff
@@ -64,7 +71,13 @@ angular.module('oppia').component('explorationSaveAndPublishButtons', {
       };
 
       $scope.getChangeListLength = function() {
-        return ChangeListService.getChangeList().length;
+        var changeListLength = ChangeListService.getChangeList().length;
+        if (changeListLength > 50) {
+          PreventPageUnloadEventService.addListener();
+        } else {
+          PreventPageUnloadEventService.removeListener();
+        }
+        return changeListLength;
```

Comment:
We need to add prevent reload here to prevent the website from reloading when there are more than 50 changes in the list and if lesser than that the listener should be removed

Reply:
But the change list is always saved in the backend, right? So why are we preventing reload?

Reply:
This was a user request. Seems that the user lost the changes once the window was closed. So once more than 50 changes are present in the list. We nudge the user to save the changes by using the prevent reload modal

Reply:
> Seems that the user lost the changes once the window was closed.

Then it looks like some other (major) issue in our workflow, are you able to reproduce it?

> We nudge the user to save the changes by using the prevent reload modal

Each change/drafts is auto-saved and the user doesn't have to do anything to save any changes.

Reply:
Hi @DubeySandeep, for context, a user actually wrote to us saying they lost their changes (they had a lot of changes that were drafts). Ideally, we don't want users to keep so many changes in their draft, I think. So we thought we should prompt users to save their changes if they have too many (we decided >50).

Reply:
> Ideally, we don't want users to keep so many changes in their draft, I think. So we thought we should prompt users to save their changes if they have too many (we decided >50).

This can be a good new feature.

>  a user actually wrote to us saying they lost their changes (they had a lot of changes that were drafts)

Sounds like a bug in the existing structure? are we able to reproduce this? Was this an automatic loss (because of some bug) or is it's because of conflicts and the user had to erase the old and outdated changes?

Reply:
I suspect it was conflicts or some schema changes were made and that resulted in invalid / incompatible draft changes. We will be fixing the save flow later this year. But we think @gp201's changes are required anyway to prevent creators from having too many draft changes.

Reply:
> I suspect it was conflicts or some schema changes were made and that resulted in invalid / incompatible draft changes. We will be fixing the save flow later this year.

Ah, I see! I was confused by @gp201 comment: `Seems that the user lost the changes once the window was closed.` Thanks for the explanation.

Instead of showing a warning (with a not very clear message) only when the user is trying to reload/leave the page, should we show a modal (with a clear message now why this is important and a call-to-action button) or warning at the bottom of the page when the user crosses a limit? [Considering the conflicts can happen while you are editing the exploration.]

Reply:
@DubeySandeep we currently use the default alert modal in the new structures editor pages to let the user know that they need to save changes. In the exp editor however, we do an "autosave" and maintain drafts so we didn't need this until now. However, given that people are tending to have large draft changes, we want to push them to manually "save" (like I mentioned earlier). Given that there is only one "save" button in the editor page, the alert message should not be too confusing. Also, we currently use this functionality in other editor pages so I think it should be fine to go ahead with @gp201's approach to reduce cases where people lose their work.

As mentioned in my prev comment, we will be working on handling conflicts and schema migration related issues later on.

Reply:
> Given that there is only one "save" button in the editor page, the alert message should not be too confusing.

But this alert message will only pop-up when users are reloading/leaving the page, right? That way it won't help the user avoid having many drafts [50+] (situations where users are adding multiple changes without reloading/leaving the page.). I believe having a separate modal will help us provide a clear warning message, call-to-action button at the correct time.

The default warning message seems correct for the new-structure project (as we don't save a draft in backend) but for the exploration cases it seems incorrect: `Changes you have made may not be saved`

![image](https://user-images.githubusercontent.com/16653571/108343364-db9d8780-7201-11eb-9fd2-48d00d6cad16.png)

Reply:
I think if creators are actively working on a page, it should be OK to allow them to work with drafts. It depends on the creators workflow. It can be a bit annoying if we keep reminding them to save their changes. We're more concerned about people leaving large amount of draft changes between sessions.

But I think you have a point regarding showing a custom modal. If we have that, it can be used across the editor pages too. The trigger to show this would still remain the same though i.e. when they leave the editor. This can be done on a separate PR though if the author prefers it.

Reply:
>  we have that, it can be used across the editor pages too. 

I'm not sure whether we need this in other places as we don't auto-save draft in all the places and we don't have such frequent version issues in all the places.

>  The trigger to show this would still remain the same though i.e. when they leave the editor.

I don't think you can show the custom modal on reload (because of security reasons.) [not sure!]. Anyway, I think we should show it once the user reaches a threshold level because having multiple changes in craft also make auto-save payloads/response heavy and I believe we should consider that the version mismatch issues can happen when the user is updating the exploration, so it's okay to show that that modal once/twice once they reach a threshold level (and in future, we are handling the version mismatch issue in a better way so this won't be a problem.)

Note: The current alert modal has a strict message which isn't correct for our use cases.

@gp201, I think you can skip the feature to asking users to save a draft (And create a separate one later) if you are finding it complicated to implement in this PR? :)

Reply:
Cool, given the info above I think it makes sense to show a custom modal to the user once they reach a threshold of draft changes. Thanks @DubeySandeep!

Reply:
@DubeySandeep I will implement the custom modal in a separate PR. The current PR has some changes that is required for another PR. 
So, shall I remove the "reload site?" modal from this PR?

Reply:
> So, shall I remove the "reload site?" modal from this PR?

Yeah, I think we should remove it because the modal has an incorrect and not very clear message.

Reply:
DONE. Should I create a separate Issue to track the custom modal?